File: 2689-rearranging-fruits/rearranging-fruits.py
class Solution:
    def minCost(self, basket1: List[int], basket2: List[int]) -> int:
        count = Counter()

        # Count frequency difference between basket1 and basket2
        for x in basket1:
            count[x] += 1
        for x in basket2:
            count[x] -= 1

        swaps = []
        for k in count:
            if count[k] % 2 != 0:
                return -1  # Not possible if any count is odd
            swaps.extend([k] * (abs(count[k]) // 2))

        swaps.sort()
        res = 0
        if swaps:
            small = min(count.keys())
            for i in range(len(swaps) // 2):
                res += min(swaps[i], 2 * small)
        return res

File: 2689-rearranging-fruits/rearranging-fruits.go
func minCost(basket1 []int, basket2 []int) int64 {
	count := make(map[int]int)

	// Count frequency difference of each fruit between basket1 and basket2
	for _, v := range basket1 {
		count[v]++
	}
	for _, v := range basket2 {
		count[v]--
	}

	swaps := []int{}
	minValue := int(^uint(0) >> 1) // Max int
	for k, v := range count {
		if v%2 != 0 {
			return -1 // Impossible if any count is odd
		}
		if k < minValue {
			minValue = k // Track smallest fruit cost
		}
		for i := 0; i < abs(v)/2; i++ {
			swaps = append(swaps, k)
		}
	}

	sort.Ints(swaps)
	var res int64 = 0
	n := len(swaps)
	for i := 0; i < n/2; i++ {
		// Choose between direct swap or two swaps via the smallest fruit
		cost := min(swaps[i], 2*minValue)
		res += int64(cost)
	}
	return res
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
